### ISSUE TITLE
[FEATURE] add method to validate expression without expanding it

### ISSUE THREAD

**Issue by ppazos** [wardle/hermes-api#1]
There is a method to expand expressions that, if the expression is wrong, throws an exception from the expression parser.

In our implementation we allow users to prevalidate expressions, before actually expanding them, which happens when a data query is e valuated, so we can grant expressions used in queries are actually valid.

We can use the same service for the expansion to actually validate the expression, and if there is an exception, then the expression is not valid. Though when the expression is valid, all the expansion work will happen, which is less performant than just parsing the expression. Also, all codes from the expansion are just discarded, because we care only about the expression validity, not the actual expansion.

It would be useful to have an isValid(String expression): boolean method in the API to support this use case, which internally avoids all the expansion processing and just does the parsing.

Thanks!

**Comment by wardle** [comment]
Thanks @ppazos 

The current expand does three steps:
1. Parse the ECL
2. Process the AST to build a Lucene query
3. Perform the query and turn the result into domain orientated results. 

At the moment the top level API doesn't provide access to the individual steps. 

I think the options are:

- You can use expandEcl and check for exception. Simple and expansion is fast generally. But it wastes the results and does needless processing. However, when I'm building queries I like to see the results - particularly when it is a complex ECL - so perhaps that is good. We could add a max hits parameter to limit results so that this is safer to use when driven by end users. 

- You can use intersectEcl. This takes a collection of concept identifiers and performs an intersection with the expansion of the ECL expression. If you pass in an empty collection, eg an empty list, it is very fast and most of the processing will be parsing rather than expansion. This is probably the best interim solution  but it uses an exception to indicate parse errors which isn't ideal and still does redundant steps to give errors in syntax. It has an advantage in that it will flag an error if the ECL cannot be expanded properly (eg an unimplemented edge case) rather than purely checking syntax and leaving errors to when the query is used for the first time. 

- I could create a new valid? function. This would solve the immediate requirement - easy to implement as it just does the parse step and not the expansion step. Downsides: can't use the parse results for a later expansion, detailed error information about what was wrong with the ECL (line/chat number and the fragment with error) is not returned, only checks syntax and not the implementation. 

- I could create a new top level parse fn that takes ECL and returns the AST or an error. This could be used for checking syntax but also then optionally used to expand using another new top level fn. This is simple but expands the top level API (needing parse, expand and a method to check whether in error state or not). Expanding the API does have downsides but perhaps worth it. 

- There are minor variations on above options eg a top level API fn to return an error, or null/nil, if there is a syntax error. Good in that returns detailed error messages but doesn't allow onward use of the parsed AST and doesn't check actually can be expanded properly. 

I'm actually inclined to add a max-hits option so that the ECL is checked for syntax and is exercised - but in a limited way. Of course the intersect option is doing this in a way.

**Comment by ppazos** [comment]
@wardle let me test the intersection.

I think access to lower level operations could be useful too.

**Comment by ppazos** [comment]
@wardle this worked:

```
try
            {
                server.intersectEcl(
                    [],
                    expression
                )
                println "valid"
            }
            catch (Exception e)
            {
                println "invalid"
            }
```

**Comment by wardle** [comment]
Glad it works. 

But it would be nice to give you the syntax error information - which the parser generates - including the line, character and fragment that is syntactically incorrect. Those data are in the exception but I don't think I'm providing a way to easily get them at the moment for non-Clojure clients.

**Comment by ppazos** [comment]
@wardle though it's not as explicit as an isValid() method, it works OK.

I've seen the exception is ExceptionInfo and it's data field returns a map with the details.

`[:index:0, :text:wrong expression, :line:1, :column:1, :reason:[[:tag::string, :expecting:(], [:tag::string, :expecting:*], [:tag::char, :expecting:[:char-range:true, :lo:49, :hi:57]], [:tag::string, :expecting:^], [:tag::string, :expecting:>], [:tag::string, :expecting:>>], [:tag::string, :expecting:>>!], [:tag::string, :expecting:>!], [:tag::string, :expecting:<], [:tag::string, :expecting:<<], [:tag::string, :expecting:<<!], [:tag::string, :expecting:<!], [:tag::string, :expecting:/*], [:tag::char, :expecting:[:char-range:true, :lo:10, :hi:10]], [:tag::char, :expecting:[:char-range:true, :lo:13, :hi:13]], [:tag::char, :expecting:[:char-range:true, :lo:9, :hi:9]], [:tag::char, :expecting:[:char-range:true, :lo:32, :hi:32]]]]`

That is for "wrong expression" as input ecl

**Comment by wardle** [comment]
The Exception thrown is actually always a type of clojure.lang.IExceptionInfo - which has a .getData method which can return a map of keys and values representing detailed error information.

**Comment by wardle** [comment]
Sorry @ppazos - our messages crossed!

**Comment by wardle** [comment]
On reflection, I can see that validation might need to occur in two phases. A very fast test as to whether an expression is valid ECL according to the grammar, and a slower phase, likely with some processing involved, that might well be more resource intensive. To support such an approach, I've added isValidEcl() to the APIs (Java and Clojure).  I also didn't like managing control flow through the use of exceptions, when invalid user entered statements may well be invalid fairly often. However, other methods to expand or intersect ECL will continue to throw exceptions if an error is identified in processing (e.g. an unsupported element)

Released in v1.2.1102
